Correctly link edges where streets which allow cars are required

flaktack opened the pull request:
### Summary

Linking `CAR` and `CAR_PARK` edges sometimes results in edges which were not car accessible being chosen, which meant that no trip plan was returned.

This fixes the issue, with a few side-effects:
 * the lifecycle of `RoutingRequests` in `RoutingWorker` is explicit using try-with-resources. A few modifications were required so that itineraries could be generated after temporary edges were removed.
 * the list of modes in the debug client is simplified to reflect how it should be specified
 * an NPE during bike-rental planning with data from OSM is fixed

### Issue

closes #3428 

### Unit tests

No changes -- linking with the modes available in UI was manually tested.

### Code style

:ballot_box_with_check: 

### Documentation

No changes.

### Changelog

None.

flaktack commented:
@gmellemstrand I haven't tested the FLEX parts, could please test that?

Explicit tests still need to be added, so that the expected functionality is documented, verified.

gmellemstrand commented:
This fails when doing flex searches:

- NPE because of missing RoutingContext at `AccessEgressRouter.java:41` when doing a direct flex search
- NPE because of missing egressMode at `RoutingWorker.java:184` when doing transit with flex access

Test graph: https://drive.google.com/file/d/1xfE6eZ8sXHJ55yjcQc29ucGUg3pyI6dK/view?usp=sharing

Tested with the built-in client.
Direct flex search: (59.66687, 10.63073) to (59.65092, 10.64043)
Transit with flex access: (59.66687, 10.63073) to (59.90753, 10.76591)

flaktack commented:
@gmellemstrand thanks for testing and the simple to reproduce test cases. :tada: 

I managed to reproduce and fix both cases:
 * the `RoutingContext` was not set within `DirectFlexRouter`, it is now set.
 * when specifying `FLEX_ACCESS` or `FLEX_EGRESS` the "other" mode also needs to be set, so I've updated `config.js` to correctly reflect that.

flaktack commented:
I've rebased this on top of #3449 and added a `StreetModeLinkingTest`, to document / verify the current behavior.

With these changes this is ready for review and then a merge once #3449 accepted.

gmellemstrand commented:
#3449 is now merged, but there still seems to be changes here that are not related to the PR itself.

flaktack commented:
I've rebased onto `dev-2.x`, which removed the commits from #3449.

leonardehrenfried commented on src/main/java/org/opentripplanner/routing/impl/StreetVertexIndex.java:
```diff
@@ -273,11 +273,11 @@ private TraverseMode getTraverseModeForLinker(RoutingRequest options, boolean en
+      // for park and ride we will start in car mode and walk to the end vertex
+      boolean parkAndRideDepart = modes.getCar() && options.parkAndRide && !endVertex;
+      boolean onlyCarAvailable = modes.getCar() && !(modes.getWalk() || modes.getBicycle());
+      if (onlyCarAvailable || parkAndRideDepart) {
```
If I understand this PR correctly, this is the actual part that changes the linking behaviour, correct?

flaktack replied:
Yes -- everything else is needed so that this is called with the correct parameters.

leonardehrenfried commented:
If you, like me, struggle with reviewing this, I've used the following view: https://github.com/opentripplanner/OpenTripPlanner/pull/3437/files?diff=split&w=1

It hides whitespaces changes and uses the split view.

leonardehrenfried reviewed (dismissed):
The code changing the linking behaviour is correct in my view. The tests are also thorough, well thought out and test for the correct things.

Therefore I will approve this, but would appreciate a second pair of eyes by @gmellemstrand.

gmellemstrand commented:
This looks good. Good use of AutoClosable to make the RoutingContext cleanup a bit more structured. I also see that you store the explicit geometry on the TemporaryPartialStreetEdge, so that we can cleanup the temporary edges before creating an itinerary.

It is also good that you have changed many of the calls to RoutingRequest -> RoutingContext.

I did a few travel searches while logging the number of edges in the graph, to verify that everything gets cleaned up.
